File: backend/app/services/trainer_service.py
from sqlalchemy.orm import Session
from typing import List, Optional
from ..db import crud
from ..db.models import User
from ..schemas.trainer import TrainerCreate, TrainerUpdate, TrainerResponse
from ..utils.email import send_welcome_email

class TrainerService:
    def __init__(self, db: Session):
        self.db = db
    
    def create_trainer(self, trainer_data: TrainerCreate) -> TrainerResponse:
        """Create a new trainer."""
        db_trainer = crud.create_user(
            db=self.db,
            email=trainer_data.email,
            name=trainer_data.name,
            password=trainer_data.password,
            role="trainer",
            avatar=trainer_data.avatar,
            department=trainer_data.department,
            experience=trainer_data.experience
        )
        
        # Send welcome email
        send_welcome_email(trainer_data.email, trainer_data.name)
        
        return TrainerResponse.model_validate(db_trainer)
    
    def get_trainer(self, trainer_id: str) -> Optional[TrainerResponse]:
        """Get trainer by ID."""
        db_trainer = crud.get_user(self.db, trainer_id)
        if db_trainer and db_trainer.role == "trainer":
            return TrainerResponse.model_validate(db_trainer)
        return None
    
    def get_trainers(self, skip: int = 0, limit: int = 100) -> List[TrainerResponse]:
        """Get all trainers."""
        db_trainers = crud.get_users(self.db, skip, limit)
        return [
            TrainerResponse.model_validate(trainer) 
            for trainer in db_trainers 
            if trainer.role == "trainer"
        ]
    
    def update_trainer(self, trainer_id: str, trainer_data: TrainerUpdate) -> Optional[TrainerResponse]:
        """Update trainer."""
        db_trainer = crud.update_user(self.db, trainer_id, **trainer_data.model_dump(exclude_unset=True))
        if db_trainer and db_trainer.role == "trainer":
            return TrainerResponse.model_validate(db_trainer)
        return None
    
    def delete_trainer(self, trainer_id: str) -> bool:
        """Delete trainer."""
        db_trainer = crud.get_user(self.db, trainer_id)
        if db_trainer and db_trainer.role == "trainer":
            return crud.delete_user(self.db, trainer_id)
        return False 

File: frontend/src/services/api/trainerService.ts
import { USE_MOCK_DATA, API_BASE_URL } from '@/config';
import { User, ApiResponse } from './types';
import { authService } from './authService';

class TrainerService {
  private getAuthHeaders(): Record<string, string> {
    const token = authService.getToken();
    return token ? { Authorization: `Bearer ${token}` } : {};
  }

  async getTrainers(): Promise<User[]> {
    if (USE_MOCK_DATA) {
      return this.getMockTrainers();
    }

    try {
      const response = await fetch(`${API_BASE_URL}/trainers`, {
        headers: this.getAuthHeaders(),
      });
      return response.ok ? await response.json() : [];
    } catch (error) {
      console.error('Get trainers error:', error);
      return [];
    }
  }

  async getTrainer(id: string): Promise<User | null> {
    if (USE_MOCK_DATA) {
      return this.getMockTrainer(id);
    }

    try {
      const response = await fetch(`${API_BASE_URL}/trainers/${id}`, {
        headers: this.getAuthHeaders(),
      });
      return response.ok ? await response.json() : null;
    } catch (error) {
      console.error('Get trainer error:', error);
      return null;
    }
  }

  async createTrainer(trainer: Partial<User> & { password: string }): Promise<User | null> {
    if (USE_MOCK_DATA) {
      return this.mockCreateTrainer(trainer);
    }

    try {
      const response = await fetch(`${API_BASE_URL}/trainers`, {
        method: 'POST',
        headers: { ...this.getAuthHeaders(), 'Content-Type': 'application/json' },
        body: JSON.stringify(trainer),
      });
      return response.ok ? await response.json() : null;
    } catch (error) {
      console.error('Create trainer error:', error);
      return null;
    }
  }

  async updateTrainer(trainer: User): Promise<User | null> {
    if (USE_MOCK_DATA) {
      return this.mockUpdateTrainer(trainer);
    }

    try {
      const response = await fetch(`${API_BASE_URL}/trainers/${trainer.id}`, {
        method: 'PUT',
        headers: { ...this.getAuthHeaders(), 'Content-Type': 'application/json' },
        body: JSON.stringify(trainer),
      });
      return response.ok ? await response.json() : null;
    } catch (error) {
      console.error('Update trainer error:', error);
      return null;
    }
  }

  async deleteTrainer(id: string): Promise<boolean> {
    if (USE_MOCK_DATA) {
      return this.mockDeleteTrainer(id);
    }

    try {
      const response = await fetch(`${API_BASE_URL}/trainers/${id}`, {
        method: 'DELETE',
        headers: this.getAuthHeaders(),
      });
      return response.ok;
    } catch (error) {
      console.error('Delete trainer error:', error);
      return false;
    }
  }

  // Mock implementations
  private getMockTrainers(): User[] {
    return [
      {
        id: '2',
        email: 'sarah.trainer@company.com',
        name: 'Sarah Johnson',
        role: 'trainer',
        department: 'Training',
        experience: 8,
        avatar: 'https://images.unsplash.com/photo-1494790108755-2616b612b786?w=150&h=150&fit=crop&crop=face'
      }
    ];
  }

  private getMockTrainer(id: string): User | null {
    return this.getMockTrainers().find(t => t.id === id) || null;
  }

  private async mockCreateTrainer(trainerData: Partial<User> & { password: string }): Promise<User> {
    return {
      id: Date.now().toString(),
      name: trainerData.name!,
      email: trainerData.email!,
      role: 'trainer',
      department: trainerData.department || '',
      experience: trainerData.experience || 0,
    };
  }

  private async mockUpdateTrainer(trainer: User): Promise<User> {
    return trainer;
  }

  private async mockDeleteTrainer(id: string): Promise<boolean> {
    return true;
  }
}

export const trainerService = new TrainerService();
export default trainerService; 